Content.Server/CartridgeLoader/Cartridges/MessagesCartridgeSystem.cs

```diff
@@ -0,0 +1,304 @@
+        {
+            if (messageEvent.Action == MessagesUiAction.ChangeChat)
+                component.ChatUid = messageEvent.UidInput;
+        }
+
+        UpdateUiState(uid, GetEntity(args.LoaderUid), component);
+    }
+
+    /// <summary>
+    /// On cart insertion, register as background process.
+    /// </summary>
+    private void OnCartInsertion(EntityUid uid, MessagesCartridgeComponent component, CartridgeAddedEvent args)
+    {
+        _cartridgeLoaderSystem.RegisterBackgroundProgram(args.Loader, uid);
+    }
+
+    /// <summary>
+    /// On cartridge activation, connect to messages network.
+    /// </summary>
+    private void OnCartActivation(EntityUid uid, MessagesCartridgeComponent component, CartridgeActivatedEvent args)
+    {
+        _deviceNetworkSystem.ConnectDevice(uid);
+    }
+
+    /// <summary>
+    /// On cartridge deactivation, disconnect from messages network.
+    /// </summary>
+    private void OnCartDeactivation(EntityUid uid, MessagesCartridgeComponent component, CartridgeDeactivatedEvent args)
+    {
+        _deviceNetworkSystem.DisconnectDevice(uid, null);
+    }
+
+    /// <summary>
+    /// React and respond to packets from the server
+    /// </summary>
+    private void OnPacketReceived(EntityUid uid, MessagesCartridgeComponent component, DeviceNetworkPacketEvent args)
+    {
+        if (!TryComp(uid, out CartridgeComponent? cartComponent))
+            return;
+        if (args.Data.TryGetValue<MessagesServerComponent>("ServerComponent", out var server))
+            component.LastServer = server;
+        if (args.Data.TryGetValue<bool>("NameQuery", out var _))
+            SendName(uid, component, cartComponent, args.SenderAddress);
+        if (args.Data.TryGetValue<MessagesMessageData>("Message", out var message) && cartComponent.LoaderUid != null)
+        {
+            if (message.ReceiverId == GetUserUid(cartComponent))
+            {
+                var name = GetName(message.SenderId, component);
+
+                var subtitleString = Loc.GetString("messages-pda-notification-header");
+
+                _cartridgeLoaderSystem.SendNotification(
+                    cartComponent.LoaderUid.Value,
+                    $"{name} {subtitleString}",
+                    message.Content);
+            }
+        }
+
+        if (HasComp<CartridgeLoaderComponent>(cartComponent.LoaderUid))
+            UpdateUiState(uid, cartComponent.LoaderUid.Value, component);
+
+    }
+
+    /// <summary>
+    /// Sends the user's name to the server cache.
+    /// </summary>
+    private void SendName(EntityUid uid, MessagesCartridgeComponent component, CartridgeComponent cartComponent, string? address)
+    {
+        string name = GetUserName(cartComponent);
+
+        var packet = new NetworkPayload()
+        {
+            ["UserId"] = GetUserUid(cartComponent),
+            ["NewName"] = GetUserName(cartComponent),
+        };
+        _deviceNetworkSystem.QueuePacket(uid, address, packet);
+    }
+
```

Comment:
Why does this need a device network packet just to update the name in the server?

Reply:
The message cartridge system shouldn't be accessing the accessing the messaging system directly at all but updating the names should probably be handled in a different way from how it is now. While still using device networking though.